Sources/GRPCCore/Transport/RetryThrottle.swift

```diff
@@ -0,0 +1,124 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// A throttle used to rate-limit retries and hedging attempts.
+///
+/// gRPC prevents servers from being overloaded by retries and hedging by using a token-based
+/// throttling mechanism at the transport level.
+///
+/// Each client transport maintains a throttle for the server it is connected to and gRPC records
+/// successful and failed RPC attempts. Successful attempts increment the number of tokens
+/// by ``tokenRatio`` and failed attempts decrement the available tokens by one. In the context
+/// of throttling, a failed attempt is one where the server terminates the RPC with a status code
+/// which is retryable or non fatal (as defined by ``RetryPolicy/retryableStatusCodes`` and
+/// ``HedgingPolicy/nonFatalStatusCodes``) or when the client receives a pushback response from
+/// the server.
+///
+/// See also [gRFC A6: client retries](https://github.com/grpc/proposal/blob/master/A6-client-retries.md).
+public struct RetryThrottle: Sendable {
+  // Note: only three figures after the decimal point from the original token ratio are used so
+  //   all computation is done a scaled number of tokens (tokens * 1000). This allows us to all
+  //   computation in integer space.
+
+  /// The number of tokens available, multiplied by 1000.
+  private let scaledTokensAvailable: LockedValueBox<Int>
+  /// The number of tokens, multiplied by 1000.
+  private let scaledTokenRatio: Int
+  /// The maximum number of tokens, multiplied by 1000.
+  private let scaledMaximumTokens: Int
+  /// The retry threshold, multiplied by 1000. If ``scaledTokensAvailable`` is above this then
+  /// retries are permitted.
+  private let scaledRetryThreshold: Int
+
+  /// Returns the throttling token ratio.
+  ///
+  /// The number of tokens held by the throttle is incremented by this value for each successful
+  /// response. In the context of throttling, a successful response is one which:
+  /// - receives metadata from the server, or
+  /// - is terminated with a non-retryable or fatal status code.
+  ///
+  /// If the response is a pushback response then it is not considered to be successful, even if
+  /// either of the preceding conditions are met.
+  public var tokenRatio: Double {
+    Double(self.scaledTokenRatio) / 1000
+  }
+
+  /// The maximum number of tokens the throttle may hold.
+  public var maximumTokens: Int {
+    self.scaledMaximumTokens / 1000
+  }
+
+  /// The number of tokens the throttle currently has.
+  ///
+  /// If this value is less than or equal to half of ``maximumTokens`` then RPCs will not be retried
+  /// and hedging will be disabled.
+  public var tokens: Double {
```

Comment:
Nit: maybe `availableTokens` is clearer?

Reply:
Hmm, could it be confused with the number of tokens which are available to get (i.e. max tokens minus current tokens)?

Reply:
Yes right... okay that's fine then, I think this is clear enough.